samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,223 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import (
+	"context"
+	crypto "crypto/rand"
+	"errors"
+	"fmt"
+	"time"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// global variable for logging
+var globalLogger Logger
+
+// RemoteSampler is an implementation of SamplingStrategy.
+type RemoteSampler struct {
+	// List of known centralized sampling rules
+	manifest *centralizedManifest
+
+	// proxy is used for getting quotas and sampling rules
+	xrayClient *xrayClient
+
+	// pollerStart, if true represents rule and target pollers are started
+	pollerStart bool
+
+	// samplingRules polling interval, default is 300 seconds
+	samplingRulesPollingInterval time.Duration
+
+	// Unique ID used by XRay service to identify this client
+	clientID string
+
+	// Provides system time
+	clock Clock
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*RemoteSampler)(nil)
+
+// NewRemoteSampler returns a centralizedSampler which decides to sample a given request or not.
+func NewRemoteSampler(ctx context.Context, opts ...Option) (*RemoteSampler, error) {
+	cfg := newConfig(opts...)
+
+	// Generate clientID
+	var r [12]byte
+
+	_, err := crypto.Read(r[:])
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate client ID: %w", err)
+	}
+
+	id := fmt.Sprintf("%02x", r)
+
+	clock := &DefaultClock{}
+
+	m := &centralizedManifest{
+		rules: []*centralizedRule{},
+		index: map[string]*centralizedRule{},
+		clock: clock,
+	}
+
+	remoteSampler := &RemoteSampler{
+		pollerStart:                  false,
+		clock:                        clock,
+		manifest:                     m,
+		clientID:                     id,
+		xrayClient:                   newClient(cfg.proxyEndpoint),
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+	}
+
+	// starts the rule poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+func (rs *RemoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	// ToDo: add business logic for remote sampling
+
+	return sdktrace.SamplingResult{}
+}
+
+func (rs *RemoteSampler) Description() string {
+	return "remote sampling with AWS X-Ray"
+}
+
+func (rs *RemoteSampler) start(ctx context.Context) {
+	if !rs.pollerStart {
+		rs.pollerStart = true
+		rs.startRulePoller(ctx)
+	}
+}
+
+func (rs *RemoteSampler) startRulePoller(ctx context.Context) {
```

Comment:
We should only have one poller for rules and targets, notably there should only be one goroutine that is fetching rules or targets based on their appropriate tick. This removes needing to think about any threadsafety issues.

Reply:
Not sure about that because we typically refresh sampling rules every 5 mins and sampling targets every 10 seconds.

Reply:
Yeah the same goroutine should be polling rules every 5 min *or* polling targets every 10s, but you shouldn't have two goroutines that may execute at the same time.

Reply:
It won't execute at a same time and even if it will then those would be 2 separate processes right? I don't want to overburden the `getSamplingRules` and `getSamplingTargets` logic in a single goroutine. I think this would look much cleaner. Also, I'm not sure if we can combine those in a single goroutine programmatically as we would require both the logic to run in an infinite for loop.

Reply:
> It won't execute at a same time

It will because the timing will overlap sometimes.

The logic including any potential cleanliness doesn't change that much I think

```
for {
  select {
    case <-rulesTicker.C():
      fetchRules()
    case <-targetsTicker.C()
      fetchTargets()
  }
}
```

This ensures the pollers are serialized onto a single thread to avoid simultaneously updating.

That being said, another thing to keep in mind is that targets fetch isn't a fixed tick, it is a value returned in the actual targets response

https://docs.aws.amazon.com/xray/latest/api/API_SamplingTargetDocument.html#xray-Type-SamplingTargetDocument-Interval

So presumably you can't use a ticker for the targets fetch, I'm not sure if there is any better pattern in Go but I guess it may involve a goroutine that sleeps and signals a channel

```
for {
  case := <-rulesTicker.C()
    fetchRules()
  case <-targetsChan
    fetchTargets()
}

fetchTargets() {
  ...
  nextFetch = min(rules.intervals)
  s.manifest = newManifest
  go func() {
    time.Sleep(nextFetch)
    targetsChan <- struct{}
}
```